### ISSUE TITLE
Make the Mpris2 Widget support actions

### ISSUE THREAD

**Issue by karthikiyengar** [taffybar/taffybar#517]
It would be great if the MPRIS2 Widget can support actions upon click. The [Mpris2 client](https://github.com/taffybar/taffybar/blob/d4a7d3d32e0d1dacd5139504d762c66cfc4ff458/src/System/Taffybar/DBus/Client/MPRIS2.hs) already has bindings for issuing [actions](https://github.com/taffybar/taffybar/blob/d4a7d3d32e0d1dacd5139504d762c66cfc4ff458/dbus-xml/org.mpris.MediaPlayer2.Player.xml). 

It should be possible to use the CanPlay, CanPause, CanGoPrevious and CanGoNext fields from the [Mpris2 interface](https://specifications.freedesktop.org/mpris-spec/2.2/Player_Interface.html#Property:CanGoNext) to determine what actions can be performed. We would need to store the previous playback state manually.

In terms of UX, we can either choose to include buttons, or we could map these actions to various click actions (e.g single click issues play/pause toggle, double click goes forward etc.) 

Gitter Discussion: https://gitter.im/taffybar/Lobby?at=6104967572c37b6f4581e687

**Comment by IvanMalison** [comment]
Related: #310

**Comment by IvanMalison** [comment]
@karthikiyengar 
With https://github.com/taffybar/taffybar/commit/a6d0a136b46a1b605e85f40af36cd7eb66a14066

widgets should now stay on screen by default when they are paused.

This should make implementing this functionality much easier.

Also, it should be possible to make totally custom widgets with the new function

**Comment by IvanMalison** [comment]
@karthikiyengar In https://github.com/taffybar/taffybar/commit/a0078f95b4324dc3647117690fcb8b144aedb8c7 I made it so that the Play/Pause command is issued whenever the widget is pressed. Basically, it should only be a matter of slapping some icons into the player grid widget to add support for other actions.

What actions would you like to see implemented to consider this issue closed?

I'd really appreaciate some help with just implementing the icon/layout stuff.

**Comment by karthikiyengar** [comment]
Hey Ivan,

Thanks a lot for this. I'd be happy to help with the icon stuff, I'll open a PR soon!

**Comment by kingyakshaver** [comment]
@karthikiyengar are you still working on this? I'm new to haskell but would maybe want to pick this up.

**Comment by karthikiyengar** [comment]
Hey @kingyakshaver , I haven't started yet, but I plan to chip on it over the weekend. I'll drop a comment if I run into brick walls, but I intend to take a crack at it first if that's alright 😀

**Comment by kingyakshaver** [comment]
> Hey @kingyakshaver , I haven't started yet, but I plan to chip on it over the weekend. I'll drop a comment if I run into brick walls, but I intend to take a crack at it first if that's alright

Yeah, of course. @IvanMalison is there something else I could help out with?

**Comment by karthikiyengar** [comment]
So I took a few hours to dive into the codebase today, and after fighting with `nix` and friends for a while, I managed to set up a primitive development environment. Due to my limited Haskell chops, most of my time was just messing around with the button adding logic and trying to understand the Haskell/GTK/Dbus universe better :sweat_smile:. 

I think we need to figure out a couple of things though:
-  In terms of UX, do we want to always show the play/forward/back buttons (when available) or should we maybe do this on hover? Any other suggestions?
- Are we planning to ship the buttons by default? I remember @IvanMalison mentioning that he wants to generalize how the widget is created. If that is the case, do we need further config driven customization for the buttons?

@kingyakshaver - I'm also pretty new to non-trivial Haskell, so progress on this one is going to be pretty slow. I'd be happy to drop this in case you want take it up. I'd also be open to pairing, so feel free to hit me up on Gitter/Matrix.

**Comment by IvanMalison** [comment]
> * In terms of UX, do we want to always show the play/forward/back buttons (when available) or should we maybe do this on hover? Any other suggestions?

I'll leave that up to you. I'm not really a UI person.



> * Are we planning to ship the buttons by default?

Yes, I think thats probably a good idea. At the very least, i think it would be nice to have some indication of whether the player is playing or paused, now that the player stays up by default. Whether to add more buttons after that is up to you.

> I remember @IvanMalison mentioning that he wants to generalize how the widget is created. If that is the case, do we need further config driven customization for the buttons?

Not necessarily, but that could be nice. At this point if you really need more customization you can always replace with the custom haskell function.

**Comment by karthikiyengar** [comment]
Right, so after chatting a bit on Gitter, here's what we decided.
- Use sensible defaults for  play/pause/forward/back/fallback icons unless someone has specific suggestions
- The icons will be visible at all times for now.
- No customization options for now
- Explore possibility of scrolling text for the title/artist text.

**Comment by IvanMalison** [comment]
> * Explore possibility of scrolling text for the title/artist text.

this part can probably just be a separate change.
